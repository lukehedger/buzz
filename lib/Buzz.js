import download from './download'
import upload from './upload'

/**
 * üêù  Buzz
 *
 * @type {Class}
 */
export default class Buzz {
  /**
   * Initialise Buzz class object with options
   *
   * @param  {Object} options                     Buzz options object
   * @param  {String} [options.schemeType='LONG'] Swarm URI scheme type
   * @param  {String} options.provider            Swarm node provider URL
   * @return {void}
   */
  constructor({ schemeType = 'LONG', ...options }) {
    this.setOptions(Object.assign({}, options, { schemeType }))
  }

  /**
   * Download content from Swarm at the given content hash
   *
   * @param  {String} hash Swarm content hash to download
   * @return {String}      Resolved content
   */
  download(hash) {
    return download(hash, this.getProvider(), this.getOptions('schemeType'))
  }

  /**
   * Convenience method for getting the Swarm node provider
   *
   * @return {String} Swarm node provider URL
   */
  getProvider() {
    return this.getOptions('provider')
  }

  /**
   * Get options item by key
   *
   * @param  {String} optionsKey Key of options item
   * @return {*}                 Value of options item
   */
  getOptions(optionsKey) {
    return this.options[optionsKey]
  }

  /**
   * Convenience method for setting the Swarm node provider
   *
   * @param  {String} provider Swarm node provider URL
   * @return {void}
   */
  setProvider(provider) {
    this.setOptions({ provider })
  }

  /**
   * Update options object with subset
   *
   * @param  {Object} newOptions Object to merge into options
   * @return {void}
   */
  setOptions(newOptions) {
    this.options = Object.assign({}, this.options, newOptions)
  }

  /**
   * Upload content to Swarm
   *
   * @param  {Buffer|String} data Data to upload
   * @return {String}             Swarm content hash
   */
  upload(data) {
    return upload(data, this.getProvider(), this.getOptions('schemeType'))
  }
}
