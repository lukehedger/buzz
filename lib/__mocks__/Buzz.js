export const download = jest.fn()
export const setOptions = jest.fn()
export const upload = jest.fn()

export default class Buzz {
  constructor({ schemeType = 'LONG', ...options }) {
    this.setOptions(Object.assign({}, options, { schemeType }))
  }

  download(hash) {
    return download(hash, this.getProvider(), this.getOptions('schemeType'))
  }

  getProvider() {
    return this.getOptions('provider')
  }

  getOptions(optionsKey) {
    return this.options[optionsKey]
  }

  setProvider(provider) {
    return this.setOptions({ provider })
  }

  setOptions(newOptions) {
    this.options = Object.assign({}, this.options, newOptions)

    return setOptions(newOptions)
  }

  upload(data) {
    return upload(data, this.getProvider(), this.getOptions('schemeType'))
  }
}
