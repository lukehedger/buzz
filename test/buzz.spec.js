import Buzz, { download, setOptions, upload } from '../lib/Buzz'

jest.mock('../lib/Buzz')

const SWARM_PROVIDER = 'http://localhost:8500'

describe('buzz :: Buzz', () => {
  let buzz

  beforeAll(() => {
    buzz = new Buzz({ provider: SWARM_PROVIDER })
  })

  it('Should initialise Buzz class object with options', () => {
    const actual = buzz

    expect(actual).toMatchSnapshot()
  })

  it('Should call download module with arguments', () => {
    const hash =
      'f070f4f87c0018a4026ff2c14895e9ef525e81c4cf4d768e1bf658e570927a13'

    buzz.download(hash)

    expect(download).toHaveBeenCalledWith(hash, SWARM_PROVIDER, 'LONG')
  })

  it('Should return provider option value', () => {
    const actual = buzz.getProvider()

    expect(actual).toEqual(SWARM_PROVIDER)
  })

  it('Should return requested option value', () => {
    const actual = buzz.getOptions('provider')

    expect(actual).toEqual(SWARM_PROVIDER)
  })

  it('Should return requested option value', () => {
    const actual = buzz.getOptions('provider')

    expect(actual).toEqual(SWARM_PROVIDER)
  })

  it('Should call setOptions method with provider option', () => {
    const actual = buzz.setProvider(SWARM_PROVIDER)

    expect(setOptions).toHaveBeenCalledWith({ provider: SWARM_PROVIDER })
  })

  it('Should update options object with new options', () => {
    buzz.setOptions({ provider: SWARM_PROVIDER })

    const actual = buzz.options

    expect(actual).toMatchSnapshot()
  })

  it('Should call upload module with arguments', () => {
    const data = 'buzz'

    buzz.upload(data)

    expect(upload).toHaveBeenCalledWith(data, SWARM_PROVIDER, 'LONG')
  })
})
