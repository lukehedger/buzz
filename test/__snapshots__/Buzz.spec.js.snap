// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buzz :: Buzz Should initialise Buzz class object with options 1`] = `
Buzz {
  "options": Object {
    "provider": "http://localhost:8500",
    "schemeType": "LONG",
  },
}
`;

exports[`buzz :: Buzz Should update options object with new options 1`] = `
Object {
  "provider": "http://localhost:8500",
  "schemeType": "LONG",
}
`;
